### PR TITLE
Add support for Field files

### PR DESCRIPTION
**Pull request by raethlo**
This PR adds a field management functionality to the CLI.

### CODE REVIEW

File: src/panoramic/cli/remote/writer.py [hunk 1]
```diff
@@ -38,8 +49,22 @@ def delete(self, actionable: Actionable):
...
+
+    def update(self, actionable: Actionable):
+        """Update data in the remote api."""
+        if isinstance(actionable, PanoField):
+            return self.update_field(actionable)
+        else:
+            self._upsert(actionable)
+
+    def _upsert(self, actionable: Actionable):
+        """Upsert data to remote api."""
```
Comment (mrknmc) [inline src/panoramic/cli/remote/writer.py:67]: I think we should use the upsert endpoint for both cases here just like with models  and datasets

File: src/panoramic/cli/paths.py [hunk 2]
```diff
@@ -32,6 +32,10 @@ def config_file(cls) -> Path:
     def scanned_dir() -> Path:
         return Path.cwd() / SystemDirectory.SCANNED.value
 
+    @staticmethod
+    def fields_dir(package: Path) -> Path:
+        return package / 'fields'
```
Comment (mrknmc) [inline src/panoramic/cli/paths.py:37]: We should introduce a new `SystemDirectory` value for the fields dir.

File: src/panoramic/cli/local/writer.py [hunk 3]
```diff
@@ -35,6 +42,8 @@ def write(self, actionable: Actionable, *, package: Optional[str] = None, file_n
             return self.write_model(actionable, package=package, file_name=file_name)
         elif isinstance(actionable, PanoVirtualDataSource):
             return self.write_data_source(actionable, package=package)
+        elif isinstance(actionable, PanoField):
+            return self.write_field(actionable, file_name=file_name)
```
Comment (mrknmc) [inline src/panoramic/cli/local/writer.py:46]: Are we missing the package argument here?

File: src/panoramic/cli/local/writer.py [hunk 4]
```diff
@@ -83,3 +92,28 @@ def delete_model(self, model: PanoModel):
...
+
+        logger.debug(f'About to write field {field.id}')
+        write_yaml(path, field.to_dict())
+
+    def delete_field(self, field: PanoField):
+        """Delete field from local filesystem."""
+        assert field.file_name is not None
+
+        if field.data_source:
+            path = Paths.fields_dir(self.cwd / field.data_source) / field.file_name
```
Comment (mrknmc) [inline src/panoramic/cli/local/writer.py:114]:
```suggestion
            # dataset-scope field
            path = Paths.fields_dir(self.cwd / field.data_source) / field.file_name
```

File: src/panoramic/cli/local/writer.py [hunk 5]
```diff
@@ -83,3 +92,28 @@ def delete_model(self, model: PanoModel):
...
+        else:
+            path = Paths.fields_dir(self.cwd) / field.file_name
```
Comment (mrknmc) [inline src/panoramic/cli/local/writer.py:116]:
```suggestion
            # company-scope field
            path = Paths.fields_dir(self.cwd) / field.file_name
```

File: src/panoramic/cli/local/writer.py [hunk 6]
```diff
@@ -83,3 +92,28 @@ def delete_model(self, model: PanoModel):
...
         logger.debug(f'About to delete model {model.id}')
         delete_file(path)
+
+    def write_field(self, field: PanoField, *, file_name: Optional[str] = None):
+        """Write model to local filesystem."""
+        if file_name is None:
+            file_name = f'{field.slug}{FileExtension.FIELD_YAML.value}'
+
+        if field.data_source:
+            path = Paths.fields_dir(self.cwd / field.data_source) / file_name
```
Comment (mrknmc) [inline src/panoramic/cli/local/writer.py:102]:
```suggestion
            # dataset-scope field
            path = Paths.fields_dir(self.cwd / field.data_source) / file_name
```

File: src/panoramic/cli/local/writer.py [hunk 7]
```diff
@@ -83,3 +92,28 @@ def delete_model(self, model: PanoModel):
...
+        else:
+            path = Paths.fields_dir(self.cwd) / file_name
```
Comment (mrknmc) [inline src/panoramic/cli/local/writer.py:104]:
```suggestion
            # company-scope field
            path = Paths.fields_dir(self.cwd) / file_name
```

File: src/panoramic/cli/local/get.py [hunk 8]
```diff
@@ -11,6 +11,8 @@ def get_state(target_dataset: Optional[str] = None) -> VirtualState:
     packages = FileReader().get_packages()
     data_sources = []
     models = []
+    fields = []
```
Comment (mrknmc) [inline src/panoramic/cli/local/get.py:14]: Is this missing loading of company-scoped field files?

File: src/panoramic/cli/field/client.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    ) -> List[Field]:
+        """Retrieve all fields in a given source and optionally a datasource."""
+        logger.debug(f'Listing fields for company: {company_slug} source: {data_source}')
+        params = {'company_slug': company_slug, 'virtual_data_source': data_source, 'offset': offset, 'limit': limit}
+        response = self.session.get(self.base_url, params=params)
+        response.raise_for_status()
+        response_json = response.json()['data']
+        return [Field.from_dict(d) for d in response_json]
+
+    def update_fields(self, company_slug: str, fields: List[Field]):
```
Comment (mrknmc) [inline src/panoramic/cli/field/client.py:108]: Like I mentioned elsewhere, we can get rid of this method and use `create_fields` only and rename it to `upsert_fields`

File: src/panoramic/cli/field/client.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        *,
+        slug: str,
+        group: str,
+        display_name: str,
+        data_type: str,
+        field_type: str,
+        description: Optional[str],
+        data_source: Optional[str],
+        calculation: Optional[str],
+        aggregation: Optional[str],
```
Comment (mrknmc) [inline src/panoramic/cli/field/client.py:25]: This should be a dictionary: https://github.com/panoramichq/diesel-service/blob/taxonomy-upsert/federated/taxonomy/models.py#L47

Comment (mrknmc) [reply src/panoramic/cli/field/client.py:25]: Also - are we missing `display_format`? https://github.com/panoramichq/diesel-service/blob/taxonomy-upsert/federated/taxonomy/models.py#L49

File: src/panoramic/cli/mapper/model.py [hunk 11]
```diff
@@ -36,7 +36,7 @@ def map_model_join_from_local(join: PanoModelJoin) -> ModelJoin:
     )
 
 
-def map_field_from_remote(field: ModelField) -> PanoModelField:
```
Comment (mrknmc) [inline src/panoramic/cli/mapper/model.py:39]: Not a big fan of the import path `src/panoramic/cli/mapper/model.py`. Can we go with `src/panoramic/cli/model_mapper.py` (preferred) or `src/panoramic/cli/model/mapper.py` ?

Comment (mrknmc) [reply src/panoramic/cli/mapper/model.py:39]: Same for `src/panoramic/cli/mapper/field.py`

**Review by mrknmc — COMMENTED** [review]
Had a look at the implementation, still have to go over tests

File: src/panoramic/cli/local/writer.py (hunk 3 above)
Comment (raethlo) [reply src/panoramic/cli/local/writer.py:46]: Yeah, didn't include it at first, forgot to add it. Good catch.

File: src/panoramic/cli/field/client.py (hunk 9 above)
Comment (raethlo) [reply src/panoramic/cli/field/client.py:108]: Yeah as agreed - let's change it back.

File: src/panoramic/cli/field/client.py (hunk 10 above)
Comment (raethlo) [reply src/panoramic/cli/field/client.py:25]: Good catch, must've forgotten those fields.

File: src/panoramic/cli/mapper/model.py (hunk 11 above)
Comment (raethlo) [reply src/panoramic/cli/mapper/model.py:39]: Sure, no strong feelings on that one.

File: src/panoramic/cli/field/client.py (hunk 9 above)
Comment (raethlo) [reply src/panoramic/cli/field/client.py]: Fixed in 547db1e.

File: src/panoramic/cli/local/writer.py (hunk 3 above)
Comment (raethlo) [reply src/panoramic/cli/local/writer.py]: Fixed in 99dcaa9.

File: src/panoramic/cli/mapper/model.py (hunk 11 above)
Comment (raethlo) [reply src/panoramic/cli/mapper/model.py]: Changed in 0e99fd9 and e3d739b.

File: src/panoramic/cli/local/get.py [hunk 12]
```diff
@@ -1,16 +1,24 @@
...
+    """
+    Build a representation of what VDS and models are on local filesystem.
     """
     packages = FileReader().get_packages()
     data_sources = []
     models = []
+    fields = []
+
+    if target_dataset is None:
+        for field, path in FileReader().get_global_fields():
```
Comment (mrknmc) [inline src/panoramic/cli/local/get.py:18]: Can we use one reader for both `get_packages` and `get_global_fields` ?

File: src/panoramic/cli/local/reader.py [hunk 13]
```diff
@@ -67,6 +74,11 @@ def get_packages(self) -> Iterable[FilePackage]:
                 name=d.name,
                 data_source_file=d / PresetFileName.DATASET_YAML.value,
                 model_files=list(d.glob(f'*{FileExtension.MODEL_YAML.value}')),
+                field_files=list(d.glob(f'**/*{FileExtension.FIELD_YAML.value}')),
```
Comment (mrknmc) [inline src/panoramic/cli/local/reader.py:77]: Shouldn't this be looking at `{SystemDirectory.FIELDS.value}/*{FileExtension.FIELD_YAML.value}` only?

File: src/panoramic/cli/remote/executor.py [hunk 14]
```diff
@@ -21,11 +21,11 @@ def __init__(self, company_name: str):
     def execute(self, action: Action):
         if action.is_creation:
             assert action.desired is not None
-            self.writer.write(action.desired)
+            self.writer.create(action.desired)
```
Comment (mrknmc) [inline src/panoramic/cli/remote/executor.py:24]: This should be `write` no?

File: src/panoramic/cli/remote/writer.py [hunk 15]
```diff
@@ -96,3 +108,28 @@ def delete_model(self, model: PanoModel):
...
+
+        remote_field = map_field_from_local(field)
+        try:
+            self.field_client.upsert_fields(company_slug=self.company_name, fields=[remote_field])
+        except requests.RequestException as e:
+            logger.debug(f'Failed to write field {field.slug} in dataset {field.data_source}', exc_info=True)
+            if e.response is not None and e.response.status_code == requests.codes.bad_request:
+                raise InvalidFieldException(e).extract_request_id(e)
+            raise FieldWriteException(field.data_source, field.slug).extract_request_id(e)
+        pass
```
Comment (mrknmc) [inline src/panoramic/cli/remote/writer.py:124]: Forgotten pass?

File: src/panoramic/cli/pano_model.py [hunk 16]
```diff
@@ -56,6 +56,96 @@ def __eq__(self, o: object) -> bool:
...
+            return (self.slug,)
+
+    def __hash__(self) -> int:
+        return hash(self.id)
+
+    def __eq__(self, o: object) -> bool:
+        if not isinstance(o, PanoField):
+            return False
+
+        return self.id == o.id
```
Comment (mrknmc) [inline src/panoramic/cli/pano_model.py:146]: We should compare all values, not just id here

File: src/panoramic/cli/pano_model.py [hunk 17]
```diff
@@ -56,6 +56,96 @@ def __eq__(self, o: object) -> bool:
...
+        slug: str,
+        group: str,
+        display_name: str,
+        data_type: str,
+        field_type: str,
+        description: Optional[str],
+        data_source: Optional[str],
+        calculation: Optional[str],
+        display_format: Optional[str],
+        aggregation: Optional[Dict[str, Any]],
```
Comment (mrknmc) [inline src/panoramic/cli/pano_model.py:76]: We should have a class defined for this, just like we do with joins etc. in `PanoModel`

File: src/panoramic/cli/field/client.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+            'field_type': self.field_type,
+            'description': self.description,
+            'calculation': self.calculation,
+            'aggregation': self.aggregation,
+            'data_source': self.data_source,
+            'display_format': self.display_format,
+        }
+
+    def __hash__(self) -> int:
+        return hash(self.to_dict())
```
Comment (mrknmc) [inline src/panoramic/cli/field/client.py:69]: Dict is not hashable so this doesn't work

File: tests/panoramic/cli/local/test_reader.py [hunk 19]
```diff
@@ -45,6 +53,18 @@ def test_file_package_read_models(tmp_path):
...
+def test_file_package_read_fields(tmp_path):
+    field_file = Paths.fields_dir(tmp_path) / f'test_field{FileExtension.FIELD_YAML.value}'
+    os.makedirs(os.path.dirname(field_file), exist_ok=True)
+
+    with field_file.open('w') as f:
+        f.write('slug: field_slug')
+
+    package = FilePackage(name='dataset1', data_source_file=Mock(), model_files=[], field_files=[field_file])
+
+    assert list(package.read_fields()) == [({'slug': 'field_slug'}, field_file)]
```
Comment (mrknmc) [inline tests/panoramic/cli/local/test_reader.py:70]: Can you add tests for `get_global_fields` ?

File: tests/panoramic/cli/local/test_writer.py [hunk 20]
```diff
@@ -33,3 +34,38 @@ def test_writer_delete_data_source(mock_delete_file, tmp_path):
...
+def test_writer_write_field(mock_write_yaml, tmp_path):
+    mock_field_company_scoped = Mock(spec=PanoField, slug='slug', data_source=None)
+    mock_field_vds_scoped = Mock(spec=PanoField, slug='slug', data_source='test_dataset')
+
+    FileWriter(cwd=tmp_path).write_field(mock_field_company_scoped)
+    FileWriter(cwd=tmp_path).write_field(mock_field_vds_scoped)
+
+    assert mock_write_yaml.mock_calls == [
+        call(
+            tmp_path / 'fields' / f'{mock_field_company_scoped.slug}{FileExtension.FIELD_YAML.value}',
```
Comment (mrknmc) [inline tests/panoramic/cli/local/test_writer.py:49]: You could use the `SystemDirectory` in these assertions

File: tests/panoramic/cli/mapper/test_model.py [hunk 21]
```diff
@@ -1,11 +1,11 @@
...
 
-from panoramic.cli.mapper import (
+from panoramic.cli.model.client import Model, ModelField
+from panoramic.cli.model_mapper import (
     map_columns_to_model,
     map_model_from_local,
     map_model_from_remote,
 )
-from panoramic.cli.model.client import Model, ModelField
 from panoramic.cli.pano_model import PanoModel, PanoModelField
```
Comment (mrknmc) [inline tests/panoramic/cli/mapper/test_model.py:9]: This file should be in `tests/panoramic/cli/test_model_mapper.py`

**Review by mrknmc — CHANGES_REQUESTED** [review]
Don't forget to update `setup.py` please

File: tests/panoramic/cli/mapper/test_model.py (hunk 21 above)
Comment (raethlo) [reply tests/panoramic/cli/mapper/test_model.py:9]: Forgot to move it thanks.

File: tests/panoramic/cli/local/test_reader.py (hunk 19 above)
Comment (raethlo) [reply tests/panoramic/cli/local/test_reader.py:70]: Should be covered by validation tests added here https://github.com/panoramichq/panoramic-cli/pull/12

File: src/panoramic/cli/field/client.py (hunk 18 above)
Comment (raethlo) [reply src/panoramic/cli/field/client.py:69]: You're right - also problem for all the other remote model classes https://github.com/panoramichq/panoramic-cli/blob/fd7bcf23cd32bffe97f3aa1cfb4678f58a872686/src/panoramic/cli/model/client.py#L139 (thought we already fixed that?)

File: src/panoramic/cli/remote/executor.py (hunk 14 above)
Comment (raethlo) [reply src/panoramic/cli/remote/executor.py]: Fixed in c594dc6, thanks.

File: src/panoramic/cli/remote/writer.py (hunk 15 above)
Comment (raethlo) [reply src/panoramic/cli/remote/writer.py]: Fixed in d6ae14fdbb23cb206b496e93be417f8a999c0f1b.

File: src/panoramic/cli/pano_model.py (hunk 16 above)
Comment (raethlo) [reply src/panoramic/cli/pano_model.py]: Fixed in 81452a00f02d4ef2d193499b8ebbd57901575602.

File: tests/panoramic/cli/mapper/test_model.py (hunk 21 above)
Comment (raethlo) [reply tests/panoramic/cli/mapper/test_model.py]: dd3be040aef4b3a787ce1945889d19de6edf7fe3

File: src/panoramic/cli/field/client.py (hunk 18 above)
Comment (raethlo) [reply src/panoramic/cli/field/client.py:69]: Let's fix it in a follow-up for all the affected classes.

Comment (mrknmc) [reply src/panoramic/cli/field/client.py:69]: So did I... ok let's create a bug ticket and fix there

File: src/panoramic/cli/local/get.py (hunk 12 above)
Comment (raethlo) [reply src/panoramic/cli/local/get.py]: Fixed in 28fa8e9f243a1f098bc26a2b952039de6079b8c6.

File: src/panoramic/cli/pano_model.py (hunk 17 above)
Comment (raethlo) [reply src/panoramic/cli/pano_model.py]: Done in 2e0af3a9380aa2ee124e37a8319df8f13e945f77.

File: src/panoramic/cli/local/reader.py (hunk 13 above)
Comment (raethlo) [reply src/panoramic/cli/local/reader.py:77]: Done in 6791f903757b3015b22c3f46a46a099f39214499.

File: src/panoramic/cli/field/client.py [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+        return cls(type=inputs['type'], params=inputs.get('params'))
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {'type': self.type, 'params': self.params}
+
+    def __hash__(self) -> int:
+        return hash(json.dumps(self.to_dict()))
+
+    def __eq__(self, o: object) -> bool:
+        if not isinstance(o, Field):
```
Comment (mrknmc) [inline src/panoramic/cli/field/client.py:30]:
```suggestion
        if not isinstance(o, Aggregation):
```

File: src/panoramic/cli/field/client.py [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+        aggregation = inputs.get('aggregation')
+        return cls(
+            slug=inputs['slug'],
+            group=inputs['group'],
+            display_name=inputs['display_name'],
+            data_type=inputs['data_type'],
+            field_type=inputs['field_type'],
+            description=inputs.get('description'),
+            data_source=inputs.get('data_source'),
+            aggregation=Aggregation.from_dict(aggregation) if aggregation else None,
```
Comment (mrknmc) [inline src/panoramic/cli/field/client.py:73]:
```suggestion
            aggregation=Aggregation.from_dict(aggregation) if aggregation is not None else None,
```
nitpick

File: src/panoramic/cli/field/client.py [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            'slug': self.slug,
+            'group': self.group,
+            'display_name': self.display_name,
+            'data_type': self.data_type,
+            'field_type': self.field_type,
+            'description': self.description,
+            'calculation': self.calculation,
+            'aggregation': self.aggregation.to_dict() if self.aggregation else None,
```
Comment (mrknmc) [inline src/panoramic/cli/field/client.py:87]:
```suggestion
            'aggregation': self.aggregation.to_dict() if self.aggregation is not None else None,
```
nitpick

File: src/panoramic/cli/pano_model.py [hunk 25]
```diff
@@ -56,6 +58,101 @@ def __eq__(self, o: object) -> bool:
...
+        aggregation = inputs.get('aggregation')
+        return cls(
+            slug=inputs['slug'],
+            group=inputs['group'],
+            display_name=inputs['display_name'],
+            data_type=inputs['data_type'],
+            field_type=inputs['field_type'],
+            description=inputs.get('description'),
+            data_source=inputs.get('data_source'),
+            aggregation=Aggregation.from_dict(aggregation) if aggregation else None,
```
Comment (mrknmc) [inline src/panoramic/cli/pano_model.py:132]:
```suggestion
            aggregation=Aggregation.from_dict(aggregation) if aggregation is not None else None,
```

**Review by mrknmc — APPROVED** [review]
Last minor touches and we're good to go 🚀

**Pull request merged**